infra/modules/service/access_logs.tf

```
@@ -0,0 +1,106 @@
+locals {
+  elb_account_map = {
+    "us-east-1" : "127311923021",
+    "us-east-2" : "033677994240",
+    "us-west-1" : "027434742980",
+    "us-west-2" : "797873946194"
+  }
+  log_file_transition = {
+    STANDARD_IA = 30
+    GLACIER     = 60
+  }
+
+  log_file_deletion = 0
+}
+
+resource "aws_s3_bucket" "access_logs" {
+  bucket_prefix = "${var.service_name}-access-logs"
+  force_destroy = false
+  # checkov:skip=CKV2_AWS_62:Event notification not necessary for this bucket expecially due to likely use of lifecycle rules
+  # checkov:skip=CKV_AWS_18:Access logging was not considered necessary for this bucket
+  # checkov:skip=CKV_AWS_144:Not considered critical to the point of cross region replication
+  # checkov:skip=CKV_AWS_300:Known issue where Checkov gets confused by multiple rules
+  # 
+}
+
+resource "aws_s3_bucket_public_access_block" "access_logs" {
+  bucket = aws_s3_bucket.access_logs.id
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+data "aws_iam_policy_document" "load_balancer_logs_put_access" {
+  statement {
+    effect = "Allow"
+    resources = [
+      aws_s3_bucket.access_logs.arn,
+      "${aws_s3_bucket.access_logs.arn}/*"
+    ]
+    actions = ["s3:PutObject"]
+
+    principals {
+      type        = "AWS"
+      identifiers = ["arn:aws:iam::${local.elb_account_map[data.aws_region.current.name]}:root"]
+    }
+  }
+}
+
+resource "aws_s3_bucket_lifecycle_configuration" "lc" {
+  count  = local.log_file_transition != [] && local.log_file_deletion != 0 ? 1 : 0
+  bucket = aws_s3_bucket.access_logs.id
+  rule {
+    id     = "StorageClass"
+    status = "Enabled"
+    dynamic "transition" {
+      for_each = local.log_file_transition
+      content {
+        days          = transition.value
+        storage_class = transition.key
+      }
+    }
```

there are only two items in the list, let's simplify this and make this static rather than dynamic to improve readability
```suggestion
    transition {
      days          = 30
      storage_class = "STANDARD_IA"
    }
    transition {
      days          = 60
      storage_class = "GLACIER"
    }
```

---

I had it as a dynamic variable as this is one of the values that is likely to be different depending on the use case. With it being dynamic, adding additional transitions or changing target storage classes is a bit more straightforward (just add an additional `$STORAGE_CLASS=$NUM_DAYS`.

---

I know I'm being nitpicky, but I always optimize for reading code rather than writing code and I think it's more straightforward to have it inline for someone reading the code. The local variable + for loop is an extra layer of abstraction, so from the perspective of someone reading the code, it'll be something like:

- aws_s3_bucket_lifecycle_configuration has two rules, one that aborts uploads after 7 days, one that does transition
- looks like the transition rule is dynamic, and is based on a variable log_file_transition
- scrolls up to the top of the page where log_file_transition is defined to see keys mapping to numbers
- scrolls down to the dynamic rule to see that the key is the storage class, and the number is the days it takes to transition to that storage class

vs. if it's inline it'd be something like

- aws_s3_bucket_lifecycle_configuration has three rules, one that aborts uploads after 7 days, one that transitions to STANDARD_IA storage class in 30 days, and one that transitions to GLACIER storage class in 60 days

and from a writing code perspective, i don't think it's more difficult to add an inline transition rule than it is to add a key-value pair (more lines, but not more complex)